src/memray/_memray.pyx

```diff
@@ -1111,3 +1419,37 @@ cdef class HighWaterMarkAggregatorTestHarness:
                 )
             )
         return ret
+
+
+cdef class AllocationLifetimeAggregatorTestHarness:
+    cdef AllocationLifetimeAggregator aggregator
+
+    def add_allocation(
+        self,
+        tid,
+        address,
+        size,
+        allocator,
+        native_frame_id,
+        frame_index,
+        native_segment_generation,
+    ):
+        cdef _Allocation allocation
+        allocation.tid = tid
+        allocation.address = address
+        allocation.size = size
+        allocation.allocator = <Allocator><int>allocator
+        allocation.native_frame_id = native_frame_id
+        allocation.frame_index = frame_index
+        allocation.native_segment_generation = native_segment_generation
+        allocation.n_allocations = 1
+        self.aggregator.addAllocation(allocation)
+
+    def capture_snapshot(self):
+        return self.aggregator.captureSnapshot()
+
+    def get_allocations(self):
+        cdef shared_ptr[RecordReader] reader
+        cdef TemporalAllocationGenerator gen = TemporalAllocationGenerator()
+        gen.setup(self.aggregator.generateIndex(), reader)
```

Comment:
Why not using a `__init__/__new__` for this?

Reply:
Because we can't pass C++ objects to `__init__` or `__cinit__` - every argument to `__init__` or `__new__` needs to be convertible to a Python object, and we don't want to do that with either of these.

Although... I was trying to avoid making a copy of this big vector, but it looks like Cython has thrown it into a local temporary anyway. I think I'll switch to a `shared_ptr` for both the index and the reader, to avoid making that useless copy.

Reply:
No, making it a smart pointer makes the code much uglier. Looks like I can just make the function take an r-value reference, and std::move the value out of Cython's temporary. Much less ugly.